### ISSUE TITLE
Selective notifications

### ISSUE THREAD

**Issue by jafri** [uuosio/ascdk#25]
I need a way to implement the following pattern

```
    [[eosio::on_notify("*::transfer")]]
    void ontransfer_token (
      const name& from,
      const name& to,
      const asset& quantity,
      const string& memo
    );

    [[eosio::on_notify("atomicassets::transfer")]]
    void ontransfer_nfts (
      const name& from,
      const name& to,
      const vector<uint64_t>& asset_ids,
      const string& memo
    );
```

I saw the notify=true, but I think we need it to be selective to discern actions called, and cover * vs exact contract

**Comment by learnforpractice** [comment]
I think there is already an example:

https://github.com/uuosio/ascdk/blob/89988d2ecca9ba97f8dac3497b5e6c89844c90e0/examples/notify/receiver.ts#L11

**Comment by learnforpractice** [comment]
Ok, It's more complicated than I think. Give me a moment to write a demo for you.

**Comment by learnforpractice** [comment]
Here is the code. It can be compiled, but I have not tested it. Tell me if it does not work.

```
import {
    print,
    Contract,
    action,
    contract,
    Name,
    readActionData,
} from "as-chain";

@packer
class Transfer {
    from!: Name
    to!: Name
    quantity!: Name
    memo!: string
}

@packer
class NFTTransfer {
    from!: Name
    to!: Name
    asset_ids!: u64[]
    memo!: string
}

@contract("hello")
class MyContract extends Contract {
    @action("transfer", notify=true)
    transfer(): void {
        let data = readActionData()
        if (this.receiver == Name.fromString("atomicassets")) {
            let t = new NFTTransfer()
            t.unpack(data)
            print(`${t.from} ${t.to} ${t.asset_ids}`)
        } else {
            let t = new Transfer()
            t.unpack(data)
            print(`${t.from} ${t.to} ${t.quantity}`)
        }
    }
}

```

**Comment by jafri** [comment]
@learnforpractice I am getting

```
ERROR TS2339: Property 'from' does not exist on type '~lib/array/Array<u8>'.

             if (data.from == this.receiver) {
```

**Comment by learnforpractice** [comment]
```
if (t.from == this.receiver) {
```

**Comment by jafri** [comment]
Cool that works, just ugly error from typescript lint

<img width="582" alt="Screen Shot 2022-03-06 at 10 07 02 PM" src="https://user-images.githubusercontent.com/13677357/156971467-e9eb36f9-7568-4595-a92f-b546b0cb8797.png">

**Comment by learnforpractice** [comment]
Yeah, the following code is more pretty. See 093fb8265b39002772eed7efe8b9471e442ff258
```
import {
    unpackActionData,
} from "as-chain";

let t = unpackActionData<NFTTransfer>()
if (t.from == this.receiver) {
```

**Comment by jafri** [comment]
@learnforpractice last error here:

<img width="487" alt="Screen Shot 2022-03-06 at 10 45 18 PM" src="https://user-images.githubusercontent.com/13677357/156975116-642fab61-70d0-4321-9983-02007e08b0e6.png">


Can we change  `@action("transfer", notify=true)` to  `@action("transfer", { notify: true })`

**Comment by learnforpractice** [comment]
I think we need to keep the implementation simple enough.

**Comment by jafri** [comment]
@learnforpractice how about just `@action("transfer", true)` then? We should aim to eliminate all errors in type-checking

**Comment by learnforpractice** [comment]
Well, that reduces the readability. But for the reason of eliminating type-checking errors, I agree with the change.

**Comment by jafri** [comment]
Would you rather true or { notify: true }?

**Comment by learnforpractice** [comment]
`{ notify: true }` in a decorator is kind of wired. `true` is fine I think.

**Comment by jafri** [comment]
@learnforpractice actually I saw something you did elsewhere with singleton,

lets do `@action("transfer", notify)`

**Comment by learnforpractice** [comment]
Yeah, that's better.

**Comment by learnforpractice** [comment]
> 

Solved by e6eb2bd8db0803a597aed24f7d0ccf596490526a
